### ISSUE TITLE
Size limit exceeded

### ISSUE THREAD

**Issue by ramusbucket** [mcavage/node-ldapjs#313]
I have a local ADLDS instance. It has more than 30,000 users. I want to get them all WITHOUT paging. But when I use following code, I just get only 20,000 users. I know from documentation that I don't have to specify sizeLimit here, as 0 (unlimited) is the default. I even tried to set sizeLimit to 50,000 but no luck. Anyone has ever encounter this behavior?

var ldap = require('ldapjs');

var client = new ldap.createClient({
    url: 'ldap://localhost:5000'
});

client.bind('CN=Admin,OU=Admins,DC=com', 'Secret123', function(err, message){
    console.log(err);
    console.log(message);
});

console.log('Connected...');

var opts = {
  filter: 'objectClass=user',
  scope: 'sub',
  attributes: ['cn', 'sn', 'givenName'],  
};

client.search('ou=test,dc=com', opts, function(err, res) {
  console.log(err);
  var count = 1;
  res.on('searchEntry', function(entry) {
    count = count + 1;
    console.log('entry: ' + count.toString()+ ': ' + JSON.stringify(entry.object));
  });
  res.on('searchReference', function(referral) {
    console.log('referral: ' + referral.uris.join());
  });
  res.on('error', function(err) {
    console.error('error: ' + err.message);
  });
  res.on('end', function(result) {
    console.log('status: ' + result.status);
  });
});

**Comment by pfmooney** [comment]
It's possible that the server has a limit for results returned in a query.

**Comment by ramusbucket** [comment]
Doesn't seem so. I wrote a c# CLI application to test it. It returns me all the 30k+ records.

**Comment by pfmooney** [comment]
It is difficult to tell what's wrong in this case, assuming the ldapjs client is not emitting `error` events of any type.  Would you be willing to collect bunyan logs at the `trace` level?  We can arrange means of transfer if you wish to keep the contents private.

**Comment by ramusbucket** [comment]
Sure. There is nothing confidential as such. I'm working on the test data before I can use it in production. Here are the bunyan logs at `trace` level 

{"name":"sample-ldapjs","hostname":"Windows","pid":5768,"level":30,"msg":"Starting...","time":"2015-10-28T03:24:51.501Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":5768,"level":30,"msg":"ldap://localhost:5000","time":"2015-10-28T03:24:51.506Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":5768,"level":30,"msg":"Connected...","time":"2015-10-28T03:24:51.511Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":5768,"level":30,"msg":"Search error: null","time":"2015-10-28T03:24:51.533Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":5768,"level":30,"msg":"error: OtherError: 80090304: LdapErr: DSID-0C0903C5, comment: AcceptSecurityContext error, data 20ee, v23f0\u0000","time":"2015-10-28T03:24:51.538Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":5768,"level":30,"msg":"message: undefined","time":"2015-10-28T03:24:51.539Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":5768,"level":50,"msg":"error: 000004DC: LdapErr: DSID-0C090724, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, v23f0\u0000","time":"2015-10-28T03:24:51.539Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":19804,"level":30,"msg":"Starting...","time":"2015-10-28T03:25:12.275Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":19804,"level":30,"msg":"ldap://localhost:5000","time":"2015-10-28T03:25:12.279Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":19804,"level":30,"msg":"Connected...","time":"2015-10-28T03:25:12.285Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":19804,"level":30,"msg":"Search error: null","time":"2015-10-28T03:25:12.306Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":19804,"level":30,"msg":"error: OtherError: 80090304: LdapErr: DSID-0C0903C5, comment: AcceptSecurityContext error, data 20ee, v23f0\u0000","time":"2015-10-28T03:25:12.310Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":19804,"level":30,"msg":"message: undefined","time":"2015-10-28T03:25:12.310Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":19804,"level":50,"msg":"error: 000004DC: LdapErr: DSID-0C090724, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, v23f0\u0000","time":"2015-10-28T03:25:12.311Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":8848,"level":30,"msg":"Starting...","time":"2015-10-28T03:26:15.355Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":8848,"level":30,"msg":"ldap://localhost:5000","time":"2015-10-28T03:26:15.359Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":8848,"level":30,"msg":"Connected...","time":"2015-10-28T03:26:15.365Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":8848,"level":30,"msg":"Search error: null","time":"2015-10-28T03:26:15.385Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":8848,"level":50,"msg":"error: 00002024: LdapErr: DSID-0C06057E, comment: No other operations may be performed on the connection while a bind is outstanding., data 0, v23f0\u0000","time":"2015-10-28T03:26:15.388Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":8848,"level":30,"msg":"error: null","time":"2015-10-28T03:26:15.432Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":8848,"level":30,"msg":"message: {\"messageID\":1,\"protocolOp\":\"LDAPResult\",\"status\":0,\"matchedDN\":\"\",\"errorMessage\":\"\",\"referrals\":[],\"controls\":[]}","time":"2015-10-28T03:26:15.433Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"Starting...","time":"2015-10-28T03:26:36.445Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"ldap://localhost:5000","time":"2015-10-28T03:26:36.450Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"Connected...","time":"2015-10-28T03:26:36.455Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"Search error: null","time":"2015-10-28T03:26:36.476Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"error: null","time":"2015-10-28T03:26:36.479Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"message: {\"messageID\":1,\"protocolOp\":\"LDAPResult\",\"status\":0,\"matchedDN\":\"\",\"errorMessage\":\"\",\"referrals\":[],\"controls\":[]}","time":"2015-10-28T03:26:36.480Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 2: {\"dn\":\"CN=EffertzJ,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"EffertzJ\"}","time":"2015-10-28T03:26:39.807Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 3: {\"dn\":\"CN=WolffG,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"WolffG\"}","time":"2015-10-28T03:26:39.809Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 4: {\"dn\":\"CN=WatersM,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"WatersM\"}","time":"2015-10-28T03:26:39.810Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 5: {\"dn\":\"CN=StiedemannF,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"StiedemannF\"}","time":"2015-10-28T03:26:39.811Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 6: {\"dn\":\"CN=WillmsV,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"WillmsV\"}","time":"2015-10-28T03:26:39.812Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 7: {\"dn\":\"CN=KossD,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"KossD\"}","time":"2015-10-28T03:26:39.812Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 8: {\"dn\":\"CN=KuhlmanG,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"KuhlmanG\"}","time":"2015-10-28T03:26:39.813Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 9: {\"dn\":\"CN=WizaT,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"WizaT\"}","time":"2015-10-28T03:26:39.814Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 10: {\"dn\":\"CN=JonesO,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"JonesO\"}","time":"2015-10-28T03:26:39.814Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 11: {\"dn\":\"CN=KonopelskiJ,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"KonopelskiJ\"}","time":"2015-10-28T03:26:39.815Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 12: {\"dn\":\"CN=WuckertM,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"WuckertM\"}","time":"2015-10-28T03:26:39.816Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 13: {\"dn\":\"CN=LeschD,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"LeschD\"}","time":"2015-10-28T03:26:39.818Z","v":0}

.
.
.
.


{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 20000: {\"dn\":\"CN=McKenzieA2,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"McKenzieA2\"}","time":"2015-10-28T03:26:46.315Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":30,"msg":"entry: 20001: {\"dn\":\"CN=BoehmC2,OU=Test,DC=com\",\"controls\":[],\"userPrincipalName\":\"BoehmC2\"}","time":"2015-10-28T03:26:46.315Z","v":0}
{"name":"sample-ldapjs","hostname":"Windows","pid":17688,"level":50,"msg":"error: Size Limit Exceeded","time":"2015-10-28T03:26:46.317Z","v":0}


I have grabbed initial lines of logs and the trail of the log file. There is nothing interesting in between, I'm just listing down all the users I get.

if look for `LdapErr` in the logs - you will find 2/3 occurrences of it. I'm not sure why I get it and still the pulls a list of users. I was wondering if you have a wait for `bind` function. I highly appreciate your help

**Comment by pfmooney** [comment]
OK, so the server is griping about sizeLimit.  I can't say for sure why this is facing different limits than the C# application you wrote.  That said, there are a few things I would suggest.  Your program, as written, is not waiting for connection or bind events to complete before performing the search.  It could be that the server is placing more stringent restrictions upon you for that reason.  While there is some queuing built into the client to make it easy for those new to the async model, it's best not to rely upon that functionality.  The `bind` operation should be executed only once the client `connect` method has fired. The `search` operation should be delayed until the `bind` successfully completes.

The client will do an implicit bind for you if you pass `bindDN` and `bindCredentials` during client construction (assuming the latest version).  You would only need to wait for the `connect` event then to be confident that the client is bound and ready for commands.

**Comment by felixfbecker** [comment]
@pfmooney so `bindDN` and `bindCredentials` are new options? Can't find them in the docs or change notes. Would that allow to maybe specify bind options in the url?

**Comment by pfmooney** [comment]
@felixfbecker Considering that those options were present in the old pooled client, they aren't strictly new.  They are present in the current client has a means to ease the transition for former pooled client users.  At this point, I'm not inclined to add any more "magic" to the process (such as extracting them from the URL).

**Comment by isakar** [comment]
Not sure if this can solve your problem, but this work for me [https://www.bountysource.com/issues/28742105-erroneous-documentation-paging-paged](url) 

I got an ldap error about the limit size, and just change 'paging' to 'paged' and now all works fine.

**Comment by jsumners** [comment]
⚠️ This issue has been locked due to age. If you have encountered a recent
problem that seems to be covered by this issue, please open a new issue.

Please include a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)
when opening a new issue.
